conwayste/src/client.rs

```diff
@@ -767,170 +923,202 @@ impl MainState {
+                    if let Some(TextInputState::EnteringText) = tf.state {
+                        tf.remove_left_of_cursor(ctx);
                     }
-                    input::InputAction::KeyRelease(keycode) => {
-                        match keycode {
-                            KeyCode::Up | KeyCode::Down | KeyCode::Left | KeyCode::Right => {
-                                self.arrow_input = (0, 0);
-                                self.menu_sys.get_controls().set_menu_key_pressed(false);
-                            }
-                            _ => {}
-                        }
+                }
+            }
+            KeyCode::Delete => {
+                if let Some(tf) = self.ui_manager.textfield_from_id(Screen::Run, WidgetID::InGamePane1ChatboxTextField) {
+                    if let Some(TextInputState::EnteringText) = tf.state {
+                        tf.remove_right_of_cursor(ctx);
+                    }
+                }
+            }
+            KeyCode::Left => {
+                if let Some(tf) = self.ui_manager.textfield_from_id(Screen::Run, WidgetID::InGamePane1ChatboxTextField) {
+                    if let Some(TextInputState::EnteringText) = tf.state {
+                        tf.move_cursor_left(ctx);
+                    }
+                }
+            },
+            KeyCode::Right => {
+                if let Some(tf) = self.ui_manager.textfield_from_id(Screen::Run, WidgetID::InGamePane1ChatboxTextField) {
+                    if let Some(TextInputState::EnteringText) = tf.state {
+                        tf.move_cursor_right(ctx);
                     }
-
-                    _ => {},
                 }
             }
+            KeyCode::Up => {},    // ?? go up and down the message stack?
+            KeyCode::Down => {},
+            KeyCode::Home => {
+                if let Some(tf) = self.ui_manager.textfield_from_id(Screen::Run, WidgetID::InGamePane1ChatboxTextField) {
+                    if let Some(TextInputState::EnteringText) = tf.state {
+                        tf.cursor_home();
+                    }
+                }
+            }
+            KeyCode::End => {
+                if let Some(tf) = self.ui_manager.textfield_from_id(Screen::Run, WidgetID::InGamePane1ChatboxTextField) {
+                    if let Some(TextInputState::EnteringText) = tf.state {
+                        tf.cursor_end();
+                    }
+                }
+            }
+            _ => {} // do nothing for now
         }
-
     }
 
     // update
-    fn update_current_screen(&mut self, ctx: &mut Context) {
+    fn update_main_menu(&mut self, ctx: &mut Context) {
         self.process_menu_inputs();
 
-        let is_direction_key_pressed = {
-            self.menu_sys.get_controls().is_menu_key_pressed()
-        };
+        let mouse_point = Point2::new(self.inputs.mouse_info.position.0 as f32, self.inputs.mouse_info.position.1 as f32);
```

Comment:
This is a code smell. Why not have `mouse_info` be a `Point2`?

Reply:
I think you meant `position` not `mouse_info`, but yes it should be a `Point2`. Smelly indeed.